crates/turbopack-ecmascript/src/typescript/resolve.rs

```diff
@@ -74,45 +73,98 @@ pub async fn read_tsconfigs(
-                        tsconfig = asset;
+                        tsconfig = *asset;
+                        continue;
                     } else {
                         TsConfigIssue {
                             severity: IssueSeverity::Error.into(),
                             source_ident: tsconfig.ident(),
-                            message: StringVc::cell("extends doesn't resolve correctly".into()),
+                            message: StringVc::cell(
+                                "extends doesn't resolve correctly".to_string(),
+                            ),
                         }
                         .cell()
                         .as_issue()
                         .emit();
-                        break;
                     }
-                } else {
-                    break;
                 }
             }
         }
+        break;
     }
     Ok(configs)
 }
 
+/// Resolves tsconfig files according to TS's implementation:
+/// https://github.com/microsoft/TypeScript/blob/611a912d/src/compiler/commandLineParser.ts#L3294-L3326
+async fn resolve_extends(
+    tsconfig: AssetVc,
+    extends: &str,
+    resolve_options: ResolveOptionsVc,
+) -> Result<AssetOptionVc> {
+    let context = tsconfig.ident().path().parent();
+    let request = RequestVc::parse_string(extends.to_string());
+
+    // TS's resolution is weird, and has special behavior for different import
+    // types.
+    let extends = match &*request.await? {
+        // TS has special behvaior for "rooted" paths (absolute paths):
+        // https://github.com/microsoft/TypeScript/blob/611a912d/src/compiler/commandLineParser.ts#L3303-L3313
+        Request::Windows { path: Pattern::Constant(path) } |
+        // Server relative is treated as absolute
+        Request::ServerRelative { path: Pattern::Constant(path) } => {
+            return resolve_extends_rooted_or_relative(context, request, resolve_options, path).await;
+        }
+
+        // TS has special behvaior for (explicitly) './' and '../', but not '.' nor '..':
+        // https://github.com/microsoft/TypeScript/blob/611a912d/src/compiler/commandLineParser.ts#L3303-L3313
+        Request::Relative {
+            path: Pattern::Constant(path),
+            ..
+        } if path.starts_with("./") || path.starts_with("../") => {
+            return resolve_extends_rooted_or_relative(context, request, resolve_options, path).await;
+        }
+
+        // An empty extends is treated as "tsconfig.json"
+        Request::Empty => {
+            "tsconfig.json".to_string()
+        }
+        // All other types are treated as module imports, and joined with
+        // "tsconfig.json". This includes "relative" imports like '.' and '..'.
+        _ => {
+            format!("{extends}/tsconfig.json")
+        }
+    };
+
+    let request = RequestVc::parse_string(extends);
+    Ok(resolve(context, request, resolve_options).first_asset())
```

Comment:
```suggestion
    // There might be multiple alternatives like
    // "some/path/node_modules/xyz/abc.json" and "some/node_modules/xyz/abc.json".
    // We only want to use the first one.
    Ok(resolve(context, request, resolve_options).first_asset())
```

keeping the comment from the old version